rfc/0001-policy-server-custom-resource.md

```diff
@@ -0,0 +1,426 @@
+into the Deployment definition of Policy Server.
+
+By using a dedicated Custom Resource to define the configuration knobs of
+Policy Server we will provide a better UX to our end user (as opposed to
+writing key-values into a ConfigMap). However, exposing configuration knobs
+will require more thinking (to find the good ergonomic) and more code to
+handle that.
+
+Let's take a look at this snippet. This exposes the contents of the `sources.yaml`
+file used by Policy Server:
+
+```yaml
+  sources:
+    insecure:
+    - insecure1.registry.foo.bar
+    - insecure2.registry.foo.bar
+    secure:
+    - name: self-signed1.registry.com
+      certificate: <base64 blob>
+    - name: self-signed2.registry.com
+      certificateFrom:
+        configMapKeyRef:
+          name: name-of-the-configmap
+          key: ca_key_name
+```
+
+The way to express self-signed registries (the `secure` section) takes
+some inspiration from how Secret/ConfigMap values are mounted into
+Pods. This is definitely more "Kubernetes native", but of course we will
+have to define custom Structs to hold this data, and write more code
+to handle them.
+
+
+## Welcome back, cert-manager?
+
+As discussed before, we must have a mutation webhook to
+process `PolicyServer` and `ClusterAdmissionPolicy`
+resources.
+
+This is extremely important, because we must ensure we have
+all the finalizers set and the user is not allowed harm himself
+(eg: rename a PolicyServer, delete the `default` PolicyServer,...).
+
+The controller itself can act as a mutation webhook, that's a common
+practice with Kubebuilder.
+
+However, the webhook endpoint must be secured with a TLS certificate, which
+has to be somehow provided.
+
+This is usually done by leveraging [cert-manager](https://cert-manager.io/),
+which must be previously installed on the cluster.
+
+Should we just bite the bullet and introduce this external dependency? After all,
+cert-manager is already mandatory to install many other Kubernetes add-ons.
+
+Maybe we can extend our helm chart to have these configuration knobs:
+
+  * Install cert-manager
+  * Do not install cert-manager, because it already exists. I think this
+    should be the recommended option for production deployments
+  * Generate the secrets via helm. This [seems doable](https://medium.com/nuvo-group-tech/move-your-certs-to-helm-4f5f61338aca).
+    We would not be able to rotate them, this is definitely something for quick evaluation
+    of Kubewarden. That's not meant to be used in production
+
+However, if we introduce the cert-manager dependency, we could also leverage that to
+create the certificates used by the Policy Server instances.
+
+
+## Kubewarden controller developer experience
+
+Right now, to develop the controller, we just run `make run` and that's it.
+
+If we bring the webhook endpoint back, the local development might become more cumbersome.
+This needs more research, but it's not a blocking point.
```

Comment:
In the spirit of the previous comment, if we put the knobs needed, the controller can do all this on its own (like our old controller did if you remember :P)

Reply:
Resolved with another note